python/ray/_private/worker.py

```diff
@@ -2837,30 +2837,62 @@ def kill(actor: "ray.actor.ActorHandle", *, no_restart: bool = True):
 
 @PublicAPI
 @client_mode_hook
-def cancel(object_ref: "ray.ObjectRef", *, force: bool = False, recursive: bool = True):
-    """Cancels a task according to the following conditions.
-
-    If the specified task is pending execution, it will not be executed. If
-    the task is currently executing, the behavior depends on the ``force``
-    flag. When ``force=False``, a KeyboardInterrupt will be raised in Python
-    and when ``force=True``, the executing task will immediately exit.
-    If the task is already finished, nothing will happen.
-
-    Only non-actor tasks can be canceled. Canceled tasks will not be
-    retried (max_retries will not be respected).
-
-    Calling ray.get on a canceled task will raise a TaskCancelledError or a
-    WorkerCrashedError if ``force=True``.
+def cancel(
+    object_ref: "ray.ObjectRef", *, force: bool = False, recursive: bool = True
+) -> None:
+    """Cancels a task.
+
+    Cancel API has a different behavior depending on if it is a remote function
+    (Task) or a remote Actor method (Actor Task).
+
+    Task:
+        If the specified Task is pending execution, it is cancelled and not
+        executed. If the Task is currently executing, the behavior depends
+        on the `force flag. When `force=False, a KeyboardInterrupt is
+        raised in Python and when `force=True, the executing Task
+        immediately exits. If the Task is already finished, nothing happens.
+
+        Cancelled Tasks aren't retried. `max_task_retries` aren't respected.
+
+        Calling ray.get on a cancelled Task raises a TaskCancelledError if
+        Task hasn't been scheduled yet, RayTaskError if the Task has been scheduled
+        and interrupted, or a WorkerCrashedError if `force=True`.
+
+        If `recursive=True`, all the child Tasks and Actor Tasks
+        are cancelled. If `force=True` and `recursive=True`, `force=True`
+        is ignored for child Actor Tasks.
+
+    Actor Task:
+        If the specified Task is pending execution, it is cancelled and not
+        executed. If the Task is currently executing, the behavior depends
+        on the execution model of an Actor. If it is a regular Actor
+        or a threaded Actor, the execution isn't cancelled.
+        Actor Tasks cannot be interrupted because Actors have
+        states. If it is an async Actor, Ray cancels a `asyncio.Task`.
+        The semantic of cancellation is equivalent to asyncio's cancellation.
+        https://docs.python.org/3/library/asyncio-task.html#task-cancellation
+        If the Task has finished, nothing happens.
+
+        Only `force=False` is allowed for an Actor Task. Otherwise, it raises
+        `TypeError`. Use `ray.kill(actor)` instead to kill an Actor.
```

Comment:
This should be `ValueError` instead of `TypeError`. Since we just introduced this with https://github.com/ray-project/ray/commit/03ae7790fb8c11a40f88c3186baf6556a2eaecdd, I believe we can still change it before the 2.7 release, right?

Reply:
Also in my testing I noticed that this is not raising an error if the task already finished. This is pretty confusing behavior I think:
```
In [3]: @ray.remote
   ...: class Actor:
   ...:     def f(self):
   ...:         print("hi")
   ...: 

In [4]: a = Actor.remote()

In [5]: x = a.f.remote()

(Actor pid=6797) hi
In [6]: ray.cancel(x, force=True)
```

Reply:
Let me try fixing it now... IIRC, this didn't work (when a task is finished, we don't have task spec anymore, and there's no way to obtain if object ref is from task or actor rn. I tried object_ref.actor_id().IsNil(), but it looks like this was buggy). Do you think this PR is mergeable when this issue is fixed, or is it something we can fix as a follow up?

Reply:
Yeah, we can fix this as a follow up. But the `ValueError` versus `TypeError` needs to be fixed before merging (my first comment) because it is irreversible without an API change.

Reply:
Btw, thanks for catching this! (little embarassing, but I learned ValueError means the invalid input first time...)